src/Libraries/Nop.Services/Directory/CurrencyService.cs

```diff
@@ -6,296 +6,282 @@
 using Nop.Core.Domain.Directory;
 using Nop.Data;
 using Nop.Services.Stores;
+using Optional;
+using Triplex.Validations;
 
-namespace Nop.Services.Directory
+namespace Nop.Services.Directory;
+
+/// <summary>
+/// Currency service
+/// </summary>
+public partial class CurrencyService : ICurrencyService
 {
+    #region Fields
+
+    private readonly CurrencySettings _currencySettings;
+    private readonly IExchangeRatePluginManager _exchangeRatePluginManager;
+    private readonly IRepository<Currency> _currencyRepository;
+    private readonly IStoreMappingService _storeMappingService;
+
+    #endregion
+
+    #region Ctor
+
+    public CurrencyService(CurrencySettings currencySettings,
+        IExchangeRatePluginManager exchangeRatePluginManager,
+        IRepository<Currency> currencyRepository,
+        IStoreMappingService storeMappingService)
+    {
+        _currencySettings = currencySettings;
+        _exchangeRatePluginManager = exchangeRatePluginManager;
+        _currencyRepository = currencyRepository;
+        _storeMappingService = storeMappingService;
+    }
+
+    #endregion
+
+    #region Methods
+
+    #region Currency
+
     /// <summary>
-    /// Currency service
+    /// Deletes currency
     /// </summary>
-    public partial class CurrencyService : ICurrencyService
+    /// <param name="currency">Currency</param>
+    /// <returns>A task that represents the asynchronous operation</returns>
+    public virtual async Task DeleteCurrencyAsync(Currency currency)
     {
-        #region Fields
+        await _currencyRepository.DeleteAsync(currency);
+    }
 
-        private readonly CurrencySettings _currencySettings;
-        private readonly IExchangeRatePluginManager _exchangeRatePluginManager;
-        private readonly IRepository<Currency> _currencyRepository;
-        private readonly IStoreMappingService _storeMappingService;
+    /// <summary>
+    /// Gets a currency
+    /// </summary>
+    /// <param name="currencyId">Currency identifier</param>
+    /// <returns>
+    /// A task that represents the asynchronous operation
+    /// The task result contains the currency
+    /// </returns>
+    public virtual async Task<Currency> GetCurrencyByIdAsync(int currencyId)
+    {
+        return await _currencyRepository.GetByIdAsync(currencyId, cache => default);
+    }
 
-        #endregion
+    /// <summary>
+    /// Gets a currency by code
+    /// </summary>
+    /// <param name="currencyCode">Currency code</param>
+    /// <returns>
+    /// A task that represents the asynchronous operation
+    /// The task result contains the currency
+    /// </returns>
+    public virtual async Task<Currency> GetCurrencyByCodeAsync(string currencyCode)
+    {
+        if (string.IsNullOrEmpty(currencyCode))
+            return null;
 
-        #region Ctor
+        return (await GetAllCurrenciesAsync(true))
+            .FirstOrDefault(c => c.CurrencyCode.ToLowerInvariant() == currencyCode.ToLowerInvariant());

```

Comment:
I think, rather than using ToLowerInvariant() you should make it 
`currencyCode.Equals(c.CurrencyCode, StringComparison.InvariantCulture)`